src/backend/columnar/columnar_customscan.c

```diff
@@ -514,38 +943,68 @@ CreateColumnarScanPath(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 	/* columnar scans are not parallel-aware, but they are parallel-safe */
 	path->parallel_safe = rel->consider_parallel;
 
+	path->param_info = get_baserel_parampathinfo(root, rel, paramRelids);
+
 	/*
-	 * We don't support pushing join clauses into the quals of a seqscan, but
-	 * it could still have required parameterization due to LATERAL refs in
-	 * its tlist.
+	 * Now, baserestrictinfo contains the clauses referencing only this rel,
+	 * and ppi_clauses (if present) represents the join clauses that reference
+	 * this rel and rels contained in paramRelids (after accounting for
+	 * ECs). Combine the two lists of clauses, extracting the actual clause
+	 * from the rinfo, and filtering out pseudoconstants and SAOPs.
 	 */
-	path->param_info = get_baserel_parampathinfo(root, rel,
-												 rel->lateral_relids);
+	List *allClauses = copyObject(rel->baserestrictinfo);
+	if (path->param_info != NULL)
+	{
+		allClauses = list_concat(allClauses, path->param_info->ppi_clauses);
+	}
 
 	/*
-	 * Add cost estimates for a columnar table scan, row count is the rows estimated by
-	 * postgres' planner.
+	 * This is the set of clauses that can be pushed down for this
+	 * parameterization (with the given paramRelids), and will be used to
+	 * construct the CustomScan plan.
 	 */
-	path->rows = rel->rows;
-	path->startup_cost = 0;
+	List *pushdownClauses = FilterPushdownClauses(root, rel, allClauses);
+
+	if (EnableColumnarQualPushdown)
+	{
+		cpath->custom_private = extract_actual_clauses(pushdownClauses, false);
+	}
+
 	int numberOfColumnsRead = bms_num_members(rte->selectedCols);
-	path->total_cost = path->startup_cost +
-					   ColumnarScanCost(rel, rte->relid, numberOfColumnsRead);
 
-	return (Path *) cpath;
+	CostColumnarScan(cpath, root, rel, rte->relid, numberOfColumnsRead,
+					 list_length(cpath->custom_private));
+
+	add_path(rel, path);
 }
 
 
 /*
- * ColumnarScanCost calculates the cost of scanning the columnar table. The cost is estimated
- * by using all stripe metadata to estimate based on the columns to read how many pages
- * need to be read.
+ * CostColumnarScan calculates the cost of scanning the columnar table. The
+ * cost is estimated by using all stripe metadata to estimate based on the
+ * columns to read how many pages need to be read.
  */
-static Cost
-ColumnarScanCost(RelOptInfo *rel, Oid relationId, int numberOfColumnsRead)
+static void
+CostColumnarScan(CustomPath *cpath, PlannerInfo *root, RelOptInfo *rel,
```

Comment:
Considering the fact that chunk skip nodes will almost act like (indeed, mostly better than) indexes, should we make the `ColumnarCustomScan` based on `ColumnarIndexScanAddTotalCost` ?

Currently, `CostColumnarScan` does the estimation only based on clause selectivity, which doesn't directly imply how much we can benefit from chunk skip nodes.
I know we already eliminate the clauses with correlation < 0.9, but what about doing something similar to what is done in `ColumnarIndexScanAddTotalCost`.

My suggestion would be:

* "Not eliminate" any clauses via `CheckVarStats`.
* Refactor `ColumnarIndexScanAddTotalCost` into two pieces such that:

```
ColumnarIndexScanAddTotalCost
{
  // 1) find column selectivity & correlation
  // 2) estimate the index path cost based on selectivity & correlation
}

// becomes ....

GetIndexSelectivityAndCorrelation
{
  // 1) find column selectivity & correlation
}
ColumnarIndexLikeScanTotalCost(varSelectivity, varCorrelation)
{
  // 2) estimate the "index-like" path cost based on params
}
```

* Estimate the cost for columnar custom scan by using `ColumnarIndexLikeScanTotalCost`

What do you think @jeff-davis ? If there is something unclear, we can have a quick call as well.

Reply:
> "Not eliminate" any clauses via CheckVarStats

I think we can multiply `varCorrelation`s for all the variables we have for `ColumnarCustomScan`, and then use the result as the ultimate correlation in `ColumnarIndexLikeScanTotalCost` for `ColumnarCustomScan`.

Reply:
Well, while `maxStripeReadCount` is equal to `estimatedRows` for index scan, it should be equal to "#totalStripes" for columnar custom scan, so I'm not sure how much we can re-use the code that computes the cost for index scan.

Reply:
First of all, I'm not sure I agree that qual pushdown is mostly better than an index. In the perfect case for qual pushdown, they will be about equal in cost; but any non-perfect case will be better served by an index. For instance, an equality predicate on a unique (but uncorrelated) column, or any kind of non-random distribution that isn't detectable as correlation specifically (which would be handled very nicely by a bitmap index scan).

Basically, I see qual pushdown as a way to avoid having an index in some cases, to preserve space or reduce ingestion cost, but if you have the index it should probably be preferred. After all: the index is exact, you retrieve what you need and nothing more (as long as it covers the important columns).

When it comes to costing, for indexes, there's a single overall correlation number regardless of the number of attributes: how closely does the index order match the table order? And the quals that don't match one of the index columns are already eliminated.

Qual pushdown is a bit different because there's no overall correlation. Options:
1. Try to track overall correlation. This doesn't sound promising, because `CREATE STATISTICS` doesn't seem to help. And even if it did, it's not likely to be a perfect match to the quals, so we'd still need a way to integrate the correlation of extra quals.
2. Try to derive some overall correlation number. If we just multiply the correlations together, a single qual on an uncorrelated column will drive the overall correlation to zero and drive the cost way up. That shouldn't happen from adding a qual. Maybe there's some other clever way, like multiplying the "complementCorrelation"s together? I'll have to think through the math on that (or let me know if you have an idea).
3. Eliminate quals on uncorrelated columns, and then `clauselist_selectivity()` should be representative of the fraction of stripes to read.

I went with #3, but I agree that #2 is appealing to share the costing code (which also makes it much more likely to choose fairly).